solidity/contracts/middleware/InterchainAccountRouter.sol

```diff
@@ -278,21 +281,41 @@ contract InterchainAccountRouter is Router {
         bytes32 _sender,
         bytes calldata _message
     ) external payable override onlyMailbox {
-        (
-            bytes32 _owner,
-            bytes32 _ism,
-            CallLib.Call[] memory _calls,
-            bytes32 _salt
-        ) = InterchainAccountMessage.decode(_message);
-
-        OwnableMulticall _interchainAccount = getDeployedInterchainAccount(
+        InterchainAccountMessage.MessageType _messageType = _message
+            .messageType();
+
+        // If the message is a reveal, we don't do any derivation of the ica address
+        // The commitment should be executed in the `verify` method of the CCIP read ISM that verified this message
+        // so here we just check that commitment -> ICA association has been cleared
+        if (_messageType == InterchainAccountMessage.MessageType.REVEAL) {
+            // The commitment should be executed in the `verify` method of the CCIP read ISM that verified this message
+            require(
+                verifiedCommitments[_message.commitment()] !=
+                    OwnableMulticall(payable(address(0))),
+                "Commitment ewas not executed"
+            );
+            return;
+        }
+
+        bytes32 _owner = _message.owner();
+        bytes32 _ism = _message.ism();
+        bytes32 _salt = _message.salt();
+
+        OwnableMulticall ica = getDeployedInterchainAccount(
             _origin,
             _owner,
             _sender,
             _ism.bytes32ToAddress(),
             _salt
         );
-        _interchainAccount.multicall{value: msg.value}(_calls);
+
+        if (_messageType == InterchainAccountMessage.MessageType.CALLS) {
+            CallLib.Call[] memory calls = _message.calls();
+            ica.multicall{value: msg.value}(calls);
+        } else {
+            bytes32 commitment = _message.commitment();
+            verifiedCommitments[commitment] = ica;
```

Comment:
what should we do if this mapping is already populated?

Reply:
what if they send the same commitment N times?
if I was a user of this API, I would expect the calls to be executed N times but as written they may only be executed once

Reply:
Fixing in another PR, as discussed offline